packages/find-root/src/index.ts

```diff
@@ -20,34 +20,56 @@ import {
  * monorepo implementations first, with tools based on custom file schemas
  * checked last.
  */
-const defaultOrder: Tool[] = [
+export const DEFAULT_TOOLS = [
   YarnTool,
   BoltTool,
   PnpmTool,
   LernaTool,
   RushTool,
+  RootTool,
 ];
 
 const isNoEntryError = (err: unknown): boolean =>
   !!err && typeof err === "object" && "code" in err && err.code === "ENOENT";
 
 export class NoPkgJsonFound extends Error {
+  directory: string;
+  constructor(directory: string) {
+    super(`No package.json could be found upwards from directory ${directory}`);
+    this.directory = directory;
+  }
+}
+
+export class NoMatchingMonorepoFound extends Error {
   directory: string;
   constructor(directory: string) {
     super(
-      `No package.json could be found upwards from the directory ${directory}`
+      `No monorepo matching the list of supported monorepos could be found upwards from directory ${directory}`
     );
     this.directory = directory;
   }
 }
 
-export async function findRoot(cwd: string): Promise<MonorepoRoot> {
+/**
+ * Given a starting folder, search that folder and its parents until a supported monorepo
+ * is found, and return a `MonorepoRoot` object with the discovered directory and a
+ * corresponding monorepo `Tool` object.
+ *
+ * By default, all predefined `Tool` implementations are included in the search -- the
+ * caller can provide a list of desired tools to restrict the types of monorepos discovered,
+ * or to provide a custom tool implementation.
```

Comment:
This sounds handy - especially the ability to provide a custom tool. I wonder though if the ability to restrict the types of discoverable monorepos isn't a YAGNI thing. OTOH, if we allow customizing a single tool, then it's really not a big problem for us to allow a list of custom tools. And at that point, "restricting" isn't really different from "custom list"

Reply:
Hah, YAGNI is definitely a concern, but I think my arguments for this approach would be:

 - I _definitely_ want to be able to provide my own custom tool from the caller.
 - I _definitely_ want to be able to specify a desired pre-baked tool.
 - But by default, I want to use a list of supported tools.
 - So... a customizable array of tools is a supserset of these requirements, but it also seems like the simplest approach that covers them all.

Reply:
👍